pulp_smash/cli.py

```diff
@@ -816,3 +819,149 @@ def _yum_apply_erratum(self, erratum):
+            self._pulp_host = smashrunner
+        else:
+            self._pulp_host = pulp_host
+
+        self._cfg = cfg
+        self._client = Client(cfg, pulp_host=self._pulp_host)
+        self._name = None
+        if raise_if_unsupported is not None:
+            self.raise_if_unsupported(*raise_if_unsupported)
+
+    @property
+    def name(self):
+        """Return the name of the Registry Client."""
+        if not self._name:
+            self._name = self._get_registry_client()
+        return self._name
+
+    def raise_if_unsupported(self, exc, message='Unsupported registry client'):
+        """Check if the registry client is supported else raise exc.
+
+        Use case::
+
+            rc = RegistryClient(cfg)
+            rc.raise_if_unsupported(unittest.SkipTest, 'Test requires podman')
+            # will raise and skip if not podman or docker
+            rc.pull('busybox')
+
+        """
+        try:
+            self.name
+        except exceptions.NoRegistryClientError:
+            raise exc(message)
+
+    def _get_registry_client(self):
+        """Talk to the host and determine the registry client.
+
+        Return "podman" or "docker" if the registry client appears to be one of
+        those.
+
+        :raises pulp_smash.exceptions.NoRegistryClientError: If unable to
+        find any valid registry client on host.
+        """
+        client = Client(self._cfg, echo_handler, pulp_host=self._pulp_host)
+        registry_clients = (
+            (('which', 'podman'), 'podman'),
+            (('which', 'docker'), 'docker'),
+        )
+        for cmd, registry_client in registry_clients:
+            if client.run(cmd).returncode == 0:
+                return registry_client
+        raise exceptions.NoRegistryClientError(
+            'Unable to determine the pregistry client used by {}. It does not '
+            'appear to be any of {}.'
+            .format(
+                self._pulp_host.hostname,
+                {rc for _, rc in registry_clients}
+            )
+        )
+
+    def _dispatch_command(self, command, *args):
+        """Dispatch a command to the registry client."""
+        # Scheme should not be part of image path, if so, remove it.
+        if args and args[0].startswith(('http://', 'https://')):
+            args = list(args)
+            args[0] = urlunsplit(
+                urlsplit(args[0])._replace(scheme='')
+            ).strip('//')
+
+        cmd = (self.name, command) + tuple(args)
+        result = self._client.run(cmd, sudo=True)
+        try:
+            # most of client responses are JSONable
+            return json.loads(result.stdout)
+        except Exception:  # pylint:disable=broad-except
+            # Python 3.4 has no specific error for json module
+            return result
+
+    pull = partialmethod(_dispatch_command, 'pull')
+    """Pulls image from registry."""
+    login = partialmethod(_dispatch_command, 'login')
```

Comment:
What credentials do we use to login?

Reply:
We do not have auth tests yet, when we add authenticated registry tests then we are going to use the `login` command, For that we will need to create an account on quay, dockerhub or any other public registry.